samples/NativeAOT/NativeAotValueManager.cs

```diff
@@ -1,10 +1,357 @@
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
 using Android.Runtime;
+using Java.Interop;
 
 namespace NativeAOT;
 
-// TODO: make a NativeAOT-specific implementation of AndroidValueManager
-// This is enough for "Hello World" to launch
-internal class NativeAotValueManager : AndroidValueManager
+internal class NativeAotValueManager : JniRuntime.JniValueManager
 {
+	readonly NativeAotTypeManager TypeManager;
+	Dictionary<int, List<IJavaPeerable>>?   RegisteredInstances = new Dictionary<int, List<IJavaPeerable>>();
 
+	public NativeAotValueManager(NativeAotTypeManager typeManager) =>
+		TypeManager = typeManager;
+
+	public override void WaitForGCBridgeProcessing ()
+	{
+	}
+
+	public override void CollectPeers ()
+	{
+		if (RegisteredInstances == null)
+			throw new ObjectDisposedException (nameof (NativeAotValueManager));
+
+		var peers = new List<IJavaPeerable> ();
+
+		lock (RegisteredInstances) {
+			foreach (var ps in RegisteredInstances.Values) {
+				foreach (var p in ps) {
+					peers.Add (p);
+				}
+			}
+			RegisteredInstances.Clear ();
+		}
+		List<Exception>? exceptions = null;
+		foreach (var peer in peers) {
+			try {
+				peer.Dispose ();
+			}
+			catch (Exception e) {
+				exceptions = exceptions ?? new List<Exception> ();
+				exceptions.Add (e);
+			}
+		}
+		if (exceptions != null)
+			throw new AggregateException ("Exceptions while collecting peers.", exceptions);
+	}
+
+	public override void AddPeer (IJavaPeerable value)
+	{
+		if (RegisteredInstances == null)
+			throw new ObjectDisposedException (nameof (NativeAotValueManager));
+
+		var r = value.PeerReference;
+		if (!r.IsValid)
+			throw new ObjectDisposedException (value.GetType ().FullName);
+		var o = PeekPeer (value.PeerReference);
+		if (o != null)
+			return;
+
+		if (r.Type != JniObjectReferenceType.Global) {
+			value.SetPeerReference (r.NewGlobalRef ());
+			JniObjectReference.Dispose (ref r, JniObjectReferenceOptions.CopyAndDispose);
+		}
+		int key = value.JniIdentityHashCode;
+		lock (RegisteredInstances) {
+			List<IJavaPeerable>? peers;
+			if (!RegisteredInstances.TryGetValue (key, out peers)) {
+				peers = new List<IJavaPeerable> () {
+					value,
+				};
+				RegisteredInstances.Add (key, peers);
+				return;
+			}
+
+			for (int i = peers.Count - 1; i >= 0; i--) {
+				var p   = peers [i];
+				if (!JniEnvironment.Types.IsSameObject (p.PeerReference, value.PeerReference))
+					continue;
+				if (Replaceable (p)) {
+					peers [i] = value;
+				} else {
+					WarnNotReplacing (key, value, p);
+				}
+				return;
+			}
+			peers.Add (value);
+		}
+	}
+
+	static bool Replaceable (IJavaPeerable peer)
+	{
+		if (peer == null)
+			return true;
+		return (peer.JniManagedPeerState & JniManagedPeerStates.Replaceable) == JniManagedPeerStates.Replaceable;
```

Comment:
Should probably use the "modern" `peer.JniManagedPeerState.HasFlag (JniManagedPeerStates.Replaceable)`.